File: src/python/meshioplusplus/vti/_vti.py
"""VTK XML ImageData, the pure-Python reference.

ImageData is a regular lattice whose geometry is three attributes -- ``Origin``,
``Spacing`` and ``WholeExtent`` -- rather than a point array. A :class:`Mesh` has
no implicit geometry, so this module expands the extent into explicit points and
one ``hexahedron`` block on read (through :mod:`meshioplusplus._grid`, the same
numbering :func:`meshioplusplus.grid` and :func:`meshioplusplus.voxelize`
produce) and derives the three attributes back from the geometry on write.

That means the writer **requires a lattice**: a mesh that is not one has no
extent to write, and a *partial* one (``voxelize``'s ``surface``/``inside``
fills, or an octree) cannot be one either, since ImageData has no way to express
a hole. Both raise :class:`WriteError` by name.

The array encoding -- base64, the 32 KiB block framing and the codecs -- is the
same VTK XML machinery :mod:`meshioplusplus.vtu` uses, and the decode path here
literally reuses ``VtuReader``'s methods rather than transcribing them.
"""

from __future__ import annotations

import base64
import xml.etree.ElementTree as ET

import numpy as np

from .._exceptions import ReadError, WriteError
from .._grid import _lattice_py, lattice_from_mesh
from .._mesh import Mesh
from .._provenance import TAG as _PROVENANCE_TAG
from ..vtu._vtu import (
    _COMPRESSION_TO_ATTR,
    VtuReader,
    _chunk_it,
    _compressor_for,
    numpy_to_vtu_type,
    vtu_to_numpy_type,
)


class _ArrayReader:
    """Just enough of a ``VtuReader`` to decode a ``<DataArray>``.

    The three decode methods are *bound from* ``VtuReader`` rather than copied:
    the base64 header framing, the per-block sizes and the byte-order handling
    are subtle enough that a second implementation would drift, and the two
    formats share the container exactly.
    """

    read_uncompressed_binary = VtuReader.read_uncompressed_binary
    read_compressed_binary = VtuReader.read_compressed_binary
    read_data = VtuReader.read_data

    def __init__(self, header_type, byte_order, compression, appended_data=None):
        self.header_type = header_type
        self.byte_order = byte_order
        self.compression = compression
        self.appended_data = appended_data


def _parse_n(text, count, dtype):
    if text is None:
        return None
    parts = text.replace(",", " ").split()
    if len(parts) < count:
        return None
    return np.array([dtype(x) for x in parts[:count]])


def read(filename):
    tree = ET.parse(str(filename))
    root = tree.getroot()
    if root.tag != "VTKFile":
        raise ReadError("Expected tag 'VTKFile'")
    if root.get("type") != "ImageData":
        raise ReadError("Expected type ImageData")

    compression = root.get("compressor")
    if compression == "vtkLZMADataCompressor":
        # The C++ reader declines lzma too; Python has the module, so this is a
        # deliberate parity choice rather than a capability gap. Removing it
        # would make the two readers accept different files.
        raise ReadError("lzma-compressed VTI is not supported")
    header_type = root.get("header_type", "UInt32")
    byte_order = root.get("byte_order")

    appended = root.find("AppendedData")
    appended_data = None
    if appended is not None:
        encoding = appended.get("encoding", "base64")
        if encoding != "base64":
            raise ReadError(f"VTI appended data encoding '{encoding}' is not supported")
        text = appended.text or ""
        appended_data = text.strip().lstrip("_")

    grid = root.find("ImageData")
    if grid is None:
        raise ReadError("No ImageData found")
    whole = _parse_n(grid.get("WholeExtent"), 6, int)
    if whole is None:
        raise ReadError("ImageData has no readable WholeExtent")
    origin = _parse_n(grid.get("Origin"), 3, float)
    if origin is None:
        origin = np.zeros(3)
    spacing = _parse_n(grid.get("Spacing"), 3, float)
    if spacing is None:
        spacing = np.ones(3)
    direction = _parse_n(grid.get("Direction"), 9, float)
    if direction is not None and not np.array_equal(direction, np.eye(3).reshape(-1)):
        raise ReadError("VTI with a non-identity Direction is not supported")

    pieces = grid.findall("Piece")
    if not pieces:
        raise ReadError("No Piece found")
    if len(pieces) > 1:
        raise ReadError("multi-piece VTI is not supported")
    piece = pieces[0]
    piece_extent = _parse_n(piece.get("Extent"), 6, int)
    if piece_extent is not None and not np.array_equal(piece_extent, whole):
        raise ReadError(
            "VTI Piece Extent differs from WholeExtent; a partial piece "
            "is not supported"
        )

    dims = np.array([whole[2 * k + 1] - whole[2 * k] for k in range(3)], dtype=np.int64)
    if np.any(dims < 0):
        raise ReadError("VTI WholeExtent is inverted")
    # A point at extent index i sits at Origin + i * Spacing, so an extent that
    # does not start at zero translates the mesh's own lo corner.
    lo = np.array([origin[k] + whole[2 * k] * spacing[k] for k in range(3)])

    points, conn = _lattice_py(dims, lo, spacing)
    cells = [] if conn is None else [("hexahedron", conn)]
    num_points = points.shape[0]
    num_cells = 0 if conn is None else conn.shape[0]

    reader = _ArrayReader(header_type, byte_order, compression, appended_data)
    point_data = {}
    cell_data = {}
    for section, sink, expected, what in (
        ("PointData", point_data, num_points, "point"),
        ("CellData", cell_data, num_cells, "cell"),
    ):
        node = piece.find(section)
        if node is None:
            continue
        for da in node.findall("DataArray"):
            name = da.get("Name")
            arr = reader.read_data(da)
            if arr.size and arr.shape[0] != expected:
                raise ReadError(
                    f"VTI {what} array '{name}' has {arr.shape[0]} rows, but the "
                    f"extent has {expected} {what}s"
                )
            if what == "cell":
                if num_cells == 0:
                    continue
                sink[name] = [arr]
            else:
                sink[name] = arr

    return Mesh(points, cells, point_data=point_data, cell_data=cell_data)


def _encode_binary(data, compression, header_type):
    """One ``<DataArray>`` body as base64, in VTK's block framing.

    Transcribed from :mod:`meshioplusplus.vtu`'s two writer closures (whose form
    is tied to its streaming-etree writer and cannot be called from here). The
    cross-check that it agrees is not this docstring: ``tests/python/test_vti.py``
    reads every encoding back through the *C++* reader and vice versa.
    """
    hdr_dtype = vtu_to_numpy_type[header_type]
    data_bytes = data.tobytes()
    if not compression:
        header = np.array(len(data_bytes), dtype=hdr_dtype)
        return base64.b64encode(header.tobytes() + data_bytes).decode()

    max_block_size = 32768
    num_blocks = -int(-len(data_bytes) // max_block_size)
    last_block_size = len(data_bytes) - (num_blocks - 1) * max_block_size
    c = _compressor_for(_COMPRESSION_TO_ATTR[compression])
    blocks = [c.compress(b) for b in _chunk_it(data_bytes, max_block_size)]
    header = np.array(
        [num_blocks, max_block_size, last_block_size] + [len(b) for b in blocks],
        dtype=hdr_dtype,
    )
    return (
        base64.b64encode(header.tobytes()).decode()
        + base64.b64encode(b"".join(blocks)).decode()
    )


def write(filename, mesh, binary=True, compression="zlib", header_type=None):
    spec = lattice_from_mesh(mesh)
    if spec is None:
        raise WriteError(
            "ImageData is a regular lattice, and this mesh is not one: it needs "
            "exactly one hexahedron block whose points tile an axis-aligned box "
            "with uniform spacing. A partial grid (voxelize's 'surface'/'inside' "
            "fill, or an octree) cannot be written as .vti either -- write it as "
            ".vtu, which stores the cells explicitly."
        )
    dims, origin, spacing = spec
    if header_type is None:
        header_type = "UInt32"
    if compression not in (None, "zlib", "lzma", "lz4", "zstd"):
        raise WriteError(f"Unknown VTI compression '{compression}'")
    if not binary:
        compression = None

    ext = f"0 {int(dims[0])} 0 {int(dims[1])} 0 {int(dims[2])}"
    lines = ['<?xml version="1.0"?>']
    attrs = 'type="ImageData" version="0.1" byte_order="LittleEndian"'
    if compression:
        attrs += f' compressor="{_COMPRESSION_TO_ATTR[compression]}"'
    if header_type != "UInt32":
        attrs += f' header_type="{header_type}"'
    lines.append(f"<VTKFile {attrs}>")
    lines.append(f"<!--{_PROVENANCE_TAG}-->")

    # %.17g, matching the C++ writer exactly: the stream default of six
    # significant digits would lose ~10 digits of a real origin, placing the grid
    # 1e-7 off its own points with nothing downstream to flag it.
    def num(v):
        return f"{float(v):.17g}"

    lines.append(
        f'<ImageData WholeExtent="{ext}" '
        f'Origin="{num(origin[0])} {num(origin[1])} {num(origin[2])}" '
        f'Spacing="{num(spacing[0])} {num(spacing[1])} {num(spacing[2])}">'
    )
    lines.append(f'<Piece Extent="{ext}">')

    def emit(section, items):
        if not items:
            return
        lines.append(f"<{section}>")
        for name, data in items:
            data = np.asarray(data)
            data = data.astype(data.dtype.newbyteorder("="), copy=False)
            vtu_type = numpy_to_vtu_type[data.dtype]
            head = f'<DataArray type="{vtu_type}" Name="{name}"'
            if data.ndim == 2:
                head += f' NumberOfComponents="{data.shape[1]}"'
            if binary:
                lines.append(head + ' format="binary">')
                lines.append(_encode_binary(data, compression, header_type))
            else:
                lines.append(head + ' format="ascii">')
                fmt = "{:.11e}" if vtu_type.startswith("Float") else "{:d}"
                lines.extend(fmt.format(v) for v in data.reshape(-1))
            lines.append("</DataArray>")
        lines.append(f"</{section}>")

    emit("PointData", sorted(mesh.point_data.items()))
    # A lattice has exactly one block, so cell_data's per-block list has one
    # entry; concatenating anyway keeps a mis-shaped input from being written
    # silently truncated.
    emit(
        "CellData",
        [
            (k, np.concatenate([np.asarray(a) for a in v]))
            for k, v in sorted(mesh.cell_data.items())
        ],
    )

    lines.append("</Piece>")
    lines.append("</ImageData>")
    lines.append("</VTKFile>")
    with open(filename, "w") as f:
        f.write("\n".join(lines) + "\n")

File: src/cpp/src/formats/vti.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// VTK XML ImageData. Reader, writer and metadata reader in one file, unlike VTU
// (whose two halves are large enough to want separating): the geometry here is
// three attributes, so all three entry points fit comfortably together.
//
// Anonymous-namespace helpers are prefixed `vti_` -- the amalgamation
// concatenates every translation unit, and `vtu_` is taken.

// System includes
#include <array>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <algorithm>
#include <sstream>
#include <string>
#include <vector>

// External includes
#include "pugixml.hpp"

// Project includes
#include "meshioplusplus/formats/vti.hpp"
#include "meshioplusplus/detail/grid_lattice.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/detail/vtk_xml.hpp"
#include "meshioplusplus/detail/provenance.hpp"
#include "meshioplusplus/detail/vtu_binary.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/parallel.hpp"

namespace meshioplusplus {

namespace {

using detail::cols;
using detail::vtu_ascii_ndarray;
using detail::vtu_type_str;

// Parse a whitespace-separated run of N numbers from an XML attribute. VTK's own
// files use plain spaces; being liberal here costs nothing and a mis-parse would
// silently relocate the whole grid.
template <class T>
bool vti_parse_n(const char* pText, T* pOut, std::size_t Count) {
    if (pText == nullptr)
        return false;
    std::istringstream is(pText);
    for (std::size_t i = 0; i < Count; ++i)
        if (!(is >> pOut[i]))
            return false;
    return true;
}

// The framing every ImageData path needs, resolved once so the mesh reader and
// the metadata reader cannot disagree about which files they accept.
struct vti_header {
    pugi::xml_node mPiece;
    detail::VtkCodec mCodec = detail::VtkCodec::None;
    std::size_t mHeaderSize = 4;
    detail::LatticeSpec mSpec;
    std::size_t mNumPoints = 0;
    std::size_t mNumCells = 0;
};

vti_header vti_parse_header(const pugi::xml_document& rDoc) {
    pugi::xml_node root = rDoc.child("VTKFile");
    if (!root)
        throw ReadError("Expected tag 'VTKFile'");
    if (std::string(root.attribute("type").as_string()) != "ImageData")
        throw ReadError("Expected type ImageData");

    vti_header h;
    const std::string compressor = root.attribute("compressor").as_string("");
    if (compressor.empty())
        h.mCodec = detail::VtkCodec::None;
    else if (compressor == detail::vtk_codec_compressor(detail::VtkCodec::Zlib))
        h.mCodec = detail::VtkCodec::Zlib;
    else if (compressor == detail::vtk_codec_compressor(detail::VtkCodec::LZ4))
        h.mCodec = detail::VtkCodec::LZ4;
    else if (compressor == detail::vtk_codec_compressor(detail::VtkCodec::ZSTD))
        h.mCodec = detail::VtkCodec::ZSTD;
    else if (compressor == detail::vtk_codec_compressor(detail::VtkCodec::LZMA))
        throw ReadError("lzma-compressed VTI not supported by the C++ reader");
    else
        throw ReadError("Unknown VTI compressor '" + compressor + "'");
    // Fail early and actionably when the file needs a codec this build lacks.
    detail::vtk_codec_require_read(h.mCodec);

    const std::string header_type = root.attribute("header_type").as_string("UInt32");
    h.mHeaderSize = (header_type == "UInt64") ? 8 : 4;

    if (root.child("AppendedData"))
        throw ReadError("appended VTI data not supported by the C++ reader");

    pugi::xml_node grid = root.child("ImageData");
    if (!grid)
        throw ReadError("No ImageData found");

    std::int64_t whole[6] = {0, 0, 0, 0, 0, 0};
    if (!vti_parse_n(grid.attribute("WholeExtent").as_string(nullptr), whole, 6))
        throw ReadError("ImageData has no readable WholeExtent");
    double origin[3] = {0.0, 0.0, 0.0};
    double spacing[3] = {1.0, 1.0, 1.0};
    if (grid.attribute("Origin"))
        vti_parse_n(grid.attribute("Origin").as_string(), origin, 3);
    if (grid.attribute("Spacing"))
        vti_parse_n(grid.attribute("Spacing").as_string(), spacing, 3);
    if (grid.attribute("Direction")) {
        // A non-identity direction matrix rotates the lattice, which an
        // axis-aligned hexahedron grid cannot express without baking the
        // rotation into the coordinates -- a different mesh from the one the
        // file describes. Refuse rather than silently drop the rotation.
        double dir[9] = {1, 0, 0, 0, 1, 0, 0, 0, 1};
        if (vti_parse_n(grid.attribute("Direction").as_string(), dir, 9)) {
            const double id[9] = {1, 0, 0, 0, 1, 0, 0, 0, 1};
            for (std::size_t i = 0; i < 9; ++i)
                if (dir[i] != id[i])
                    throw ReadError(
                        "VTI with a non-identity Direction is not supported by the "
                        "C++ reader");
        }
    }

    h.mPiece = grid.child("Piece");
    if (!h.mPiece)
        throw ReadError("No Piece found");
    if (h.mPiece.next_sibling("Piece"))
        throw ReadError("multi-piece VTI not supported by the C++ reader");
    // The piece's own extent may legally be a sub-box of the whole extent; the
    // arrays are then sized to the PIECE, and reading them against the whole
    // extent would be silently misaligned. One piece covering everything is what
    // every writer emits, so decline the rest by name.
    if (h.mPiece.attribute("Extent")) {
        std::int64_t piece[6] = {0, 0, 0, 0, 0, 0};
        if (vti_parse_n(h.mPiece.attribute("Extent").as_string(), piece, 6))
            for (std::size_t i = 0; i < 6; ++i)
                if (piece[i] != whole[i])
                    throw ReadError(
                        "VTI Piece Extent differs from WholeExtent; a partial piece "
                        "is not supported by the C++ reader");
    }

    for (std::size_t k = 0; k < 3; ++k) {
        const std::int64_t n = whole[2 * k + 1] - whole[2 * k];
        if (n < 0)
            throw ReadError("VTI WholeExtent is inverted on axis " + std::to_string(k));
        h.mSpec.mDims[k] = n;
        // The extent may start away from zero; the point at extent index i sits
        // at Origin + i * Spacing, so the mesh's own lo corner is offset by the
        // extent's start. Dropping that offset would translate the whole grid.
        h.mSpec.mOrigin[k] = origin[k] + static_cast<double>(whole[2 * k]) * spacing[k];
        h.mSpec.mSpacing[k] = spacing[k];
    }
    h.mNumPoints = static_cast<std::size_t>((h.mSpec.mDims[0] + 1) * (h.mSpec.mDims[1] + 1) *
                                            (h.mSpec.mDims[2] + 1));
    h.mNumCells = static_cast<std::size_t>(detail::lattice_num_cells(h.mSpec));
    return h;
}

NDArray vti_read_data_array(const pugi::xml_node& rDa, detail::VtkCodec codec, std::size_t hsz,
                            int& rNumComponents) {
    const std::string fmt = rDa.attribute("format").as_string("ascii");
    const DType dt = detail::dtype_from_vtu(rDa.attribute("type").as_string());
    rNumComponents = rDa.attribute("NumberOfComponents").as_int(0);
    if (fmt == "ascii")
        return detail::vtu_parse_ascii(rDa.text().get(), dt);
    if (fmt == "binary")
        return detail::vtu_parse_binary(detail::vtu_strip(rDa.text().get()), dt, codec, hsz);
    throw ReadError("VTI '" + fmt + "' data is not supported by the C++ reader");
}

// One geometry attribute value. `%.17g` rather than the stream's default six
// significant digits, which would lose ~10 digits of a real origin -- a grid
// placed 1e-7 off its own points, which nothing downstream would flag. 17 is the
// round-trip width for a double, and the identical spelling is what the numpy
// twin uses, so the two writers' attributes agree character for character.
std::string vti_num(double Value) {
    char buf[40];
    std::snprintf(buf, sizeof(buf), "%.17g", Value);
    return buf;
}

std::vector<std::string> vti_array_names(const pugi::xml_node& rPiece, const char* pSection) {
    std::vector<std::string> names;
    for (pugi::xml_node da : rPiece.child(pSection).children("DataArray"))
        names.emplace_back(da.attribute("Name").as_string());
    // The uniform mesh API hands back sorted names; match it so a summary and a
    // real read report data arrays in the same order.
    std::sort(names.begin(), names.end());
    return names;
}

}  // namespace

void write_vti(const std::string& rPath, const Mesh& rMesh, bool binary, bool zlib) {
    write_vti_codec(rPath, rMesh, binary, zlib ? detail::VtkCodec::Zlib : detail::VtkCodec::None);
}

void write_vti_codec(const std::string& rPath, const Mesh& rMesh, bool binary,
                     detail::VtkCodec codec) {
    detail::LatticeSpec spec;
    if (!detail::lattice_from_mesh(rMesh, spec))
        throw WriteError(
            "ImageData is a regular lattice, and this mesh is not one: it needs exactly one "
            "hexahedron block whose points tile an axis-aligned box with uniform spacing. A "
            "partial grid (voxelize's 'surface'/'inside' fill, or an octree) cannot be written "
            "as .vti either -- write it as .vtu, which stores the cells explicitly.");
    if (binary && codec != detail::VtkCodec::None)
        detail::vtk_codec_require_write(codec);

    std::ofstream os(rPath, std::ios::binary);
    if (!os)
        throw WriteError("Could not open file for writing: " + rPath);

    const char* fmt = binary ? "binary" : "ascii";
    auto da_header = [&](const char* type, const std::string& name, int ncomp) {
        os << "<DataArray type=\"" << type << "\" Name=\"" << name << "\"";
        if (ncomp > 0)
            os << " NumberOfComponents=\"" << ncomp << "\"";
        os << " format=\"" << fmt << "\">\n";
    };
    auto emit_bin = [&](const unsigned char* d, std::size_t n) {
        os << detail::vtu_encode_binary(d, n, binary ? codec : detail::VtkCodec::None) << "\n";
    };

    std::ostringstream ext;
    ext << "0 " << spec.mDims[0] << " 0 " << spec.mDims[1] << " 0 " << spec.mDims[2];

    os << "<?xml version=\"1.0\"?>\n";
    os << "<VTKFile type=\"ImageData\" version=\"0.1\" byte_order=\"LittleEndian\"";
    if (binary && codec != detail::VtkCodec::None)
        os << " compressor=\"" << detail::vtk_codec_compressor(codec) << "\"";
    os << ">\n";
    os << detail::provenance_render_xml_comment(detail::SlotTier::Block) << "\n";
    // Origin/Spacing/WholeExtent ARE the geometry: no Points section exists, and
    // that is the whole reason this format is worth having for a grid.
    os << "<ImageData WholeExtent=\"" << ext.str() << "\" Origin=\"" << vti_num(spec.mOrigin[0])
       << " " << vti_num(spec.mOrigin[1]) << " " << vti_num(spec.mOrigin[2]) << "\" Spacing=\""
       << vti_num(spec.mSpacing[0]) << " " << vti_num(spec.mSpacing[1]) << " "
       << vti_num(spec.mSpacing[2]) << "\">\n";
    os << "<Piece Extent=\"" << ext.str() << "\">\n";

    if (rMesh.NumPointData() != 0) {
        os << "<PointData>\n";
        for (const auto& name : rMesh.PointDataNames()) {
            const NDArray& d = rMesh.PointData(name);
            const int ncomp = (d.Shape().size() == 2) ? static_cast<int>(cols(d)) : 0;
            da_header(vtu_type_str(d.Dtype()), name, ncomp);
            if (binary)
                emit_bin(reinterpret_cast<const unsigned char*>(d.Data()), d.Nbytes());
            else
                vtu_ascii_ndarray(os, d);
            os << "</DataArray>\n";
        }
        os << "</PointData>\n";
    }

    if (rMesh.NumCellData() != 0) {
        os << "<CellData>\n";
        for (const auto& name : rMesh.CellDataNames()) {
            const std::size_t nblocks = rMesh.CellDataNumBlocks(name);
            if (nblocks == 0)
                continue;
            // A lattice has exactly one block, so there is nothing to
            // concatenate -- but iterate anyway rather than assume, since an
            // array that does not cover the block is a caller error worth not
            // writing silently truncated.
            const NDArray& first = rMesh.CellData(name, 0);
            const int ncomp = (first.Shape().size() == 2) ? static_cast<int>(cols(first)) : 0;
            da_header(vtu_type_str(first.Dtype()), name, ncomp);
            if (binary) {
                std::vector<unsigned char> buf;
                for (std::size_t bi = 0; bi < nblocks; ++bi) {
                    const NDArray& blk = rMesh.CellData(name, bi);
                    const auto* p = reinterpret_cast<const unsigned char*>(blk.Data());
                    buf.insert(buf.end(), p, p + blk.Nbytes());
                }
                emit_bin(buf.data(), buf.size());
            } else {
                for (std::size_t bi = 0; bi < nblocks; ++bi)
                    vtu_ascii_ndarray(os, rMesh.CellData(name, bi));
            }
            os << "</DataArray>\n";
        }
        os << "</CellData>\n";
    }

    os << "</Piece>\n</ImageData>\n</VTKFile>\n";
}

Mesh read_vti(const std::string& rPath, const ReadOptions& rOpts) {
    pugi::xml_document doc;
    const pugi::xml_parse_result res = doc.load_file(rPath.c_str());
    if (!res)
        throw ReadError(std::string("VTI XML parse failed: ") + res.description());

    const vti_header h = vti_parse_header(doc);

    // The extent is expanded into explicit points and hexahedra through the same
    // helper `grid()` and `voxelize()` use, so a .vti read and a grid() call of
    // the same shape produce byte-identical meshes.
    Mesh mesh = detail::lattice_build_mesh(h.mSpec);
    if (!rOpts.WantsAnyData())
        return mesh;

    for (pugi::xml_node da : h.mPiece.child("PointData").children("DataArray")) {
        const std::string name = da.attribute("Name").as_string();
        if (!rOpts.WantsArray(name))
            continue;
        int nc = 0;
        NDArray arr = vti_read_data_array(da, h.mCodec, h.mHeaderSize, nc);
        if (nc > 1)
            arr.Reshape({arr.Size() / static_cast<std::size_t>(nc), static_cast<std::size_t>(nc)});
        if (arr.Size() != 0 && detail::rows(arr) != h.mNumPoints)
            throw ReadError("VTI point array '" + name + "' has " +
                            std::to_string(detail::rows(arr)) + " rows, but the extent has " +
                            std::to_string(h.mNumPoints) + " points");
        mesh.AddPointData(name, std::move(arr));
    }
    for (pugi::xml_node da : h.mPiece.child("CellData").children("DataArray")) {
        const std::string name = da.attribute("Name").as_string();
        if (!rOpts.WantsArray(name))
            continue;
        int nc = 0;
        NDArray arr = vti_read_data_array(da, h.mCodec, h.mHeaderSize, nc);
        if (nc > 1)
            arr.Reshape({arr.Size() / static_cast<std::size_t>(nc), static_cast<std::size_t>(nc)});
        if (arr.Size() != 0 && detail::rows(arr) != h.mNumCells)
            throw ReadError("VTI cell array '" + name + "' has " +
                            std::to_string(detail::rows(arr)) + " rows, but the extent has " +
                            std::to_string(h.mNumCells) + " cells");
        if (h.mNumCells == 0)
            continue;  // no cell block to attach it to
        std::vector<NDArray> blocks;
        blocks.push_back(std::move(arr));
        mesh.AddCellData(name, std::move(blocks));
    }
    return mesh;
}

MeshMetadata read_vti_metadata(const std::string& rPath, const ReadOptions&) {
    pugi::xml_document doc;
    // parse_minimal skips escape expansion over the base64 bodies. Unlike VTU's
    // metadata path this decodes NOTHING at all: the extent attribute alone
    // gives both counts.
    const pugi::xml_parse_result res = doc.load_file(rPath.c_str(), pugi::parse_minimal);
    if (!res)
        throw ReadError(std::string("VTI XML parse failed: ") + res.description());

    const vti_header h = vti_parse_header(doc);

    MeshMetadata meta;
    meta.mNumPoints = h.mNumPoints;
    meta.mPointDim = 3;
    if (h.mNumCells != 0) {
        CellBlockInfo info;
        info.mType = "hexahedron";
        info.mNumCells = h.mNumCells;
        info.mNodesPerCell = 8;
        info.mRagged = false;
        meta.mCellBlocks.push_back(std::move(info));
    }
    meta.mPointDataNames = vti_array_names(h.mPiece, "PointData");
    meta.mCellDataNames = vti_array_names(h.mPiece, "CellData");

    // The bounding box IS the extent here, so unlike every other native metadata
    // path this one can report it for free rather than declining.
    meta.mHasBBox = true;
    for (std::size_t k = 0; k < 3; ++k) {
        meta.mBBoxMin[k] = h.mSpec.mOrigin[k];
        meta.mBBoxMax[k] =
            h.mSpec.mOrigin[k] + static_cast<double>(h.mSpec.mDims[k]) * h.mSpec.mSpacing[k];
    }
    return meta;
}

}  // namespace meshioplusplus
